### PR TITLE
Add config schema validation

### PR DESCRIPTION
**Pull request by d2r**
For issue #581

This pull request supersedes #582

This is a prototype implementation to continue along the path suggested in this comment https://github.com/nathanmarz/storm/pull/582#issuecomment-18880232

Adds a schema to each config, and validates on reading the config yaml.

Just looking for comments and direction at this point.

TODO:
- There should be a single validation function: `validate-configs-with-schemas` should just call this function and not have to test for `nil` or `instanceof FieldValidator`.  The mapping created in `b.s.config` should map the field value to this function instead of to the schema.
- We should validate on `findAndReadConfigFile`, and not on `readStormConfig`, since it will unnecessarily validate the defaults.yaml.  (It is currently there for testing.)
- Some of the types are not sufficient.  Some cases of `java.lang.Number` might need to be `Integer` instead
- Some reorganization of code (Do we want `FieldValidator`s in `Config`?)
- Lots of code intended for debugging needs to be removed.

### CODE REVIEW

File: storm-core/src/clj/backtype/storm/config.clj [hunk 1]
```diff
@@ -62,11 +80,40 @@
 (defn read-default-config []
   (clojurify-structure (Utils/readDefaultConfig)))
 
+(defn- validate-configs-with-schemas [conf]
+  (doseq [[k v] conf
+         :let [_ (println (str "trying to validate" [k v]))
```
Comment (nathanmarz) [inline storm-core/src/clj/backtype/storm/config.clj]: These printlns should be removed

File: storm-core/src/clj/backtype/storm/config.clj [hunk 2]
```diff
@@ -62,11 +80,40 @@
...
+               _ (println (str "schema is (" (get CONFIG-SCHEMA-MAP k) ")"))
+               schema (get CONFIG-SCHEMA-MAP k)]]
+    (if (instance? Config$FieldValidator schema)
+      (if (and (not (nil? v))
+               (not (.validateField schema v)))
+        (throw (IllegalArgumentException.
+                 (str "'" k "' " (.getCriteriaPredicate schema))))
+        (println (str "OK: " k " " (.getCriteriaPredicate schema)))
```
Comment (nathanmarz) [inline storm-core/src/clj/backtype/storm/config.clj]: This println should be removed

File: storm-core/src/clj/backtype/storm/config.clj [hunk 3]
```diff
@@ -25,6 +25,24 @@
          (.get f nil)
          ))
 
+;; Create a mapping of config-string -> validator
+;; Config fields must have a _SCHEMA field defined
+(def CONFIG-SCHEMA-MAP
+  (reduce (fn [mp [k v]] (assoc mp k v)) {}
```
Comment (nathanmarz) [inline storm-core/src/clj/backtype/storm/config.clj]: More elegant to do something like:

``` clojure
(->> (.getFields Config)
       (filter #(not (re-matches #".*_SCHEMA$" (.getName %))))
       (map (fn [s] [s (get-field-validator (.getField Config (str s "_SCHEMA"))])
       (into {}))
```

And get-field-validator can be a multi-method that transforms classes into an appropriate field validator, throws an error on nil, and if the value is already a field validator returns it.

```
```

Comment (nathanmarz) [inline storm-core/src/clj/backtype/storm/config.clj]: This also allows you to greatly simplify the validation code below.

**Comment by nathanmarz** [comment]
Made some inline comments. Please take note especially of the one about restructuring the code to greatly simplify it and make it more extensible at the same time to boot. Let's also get rid of all the printlns.

**Comment by d2r** [comment]
Agreed.  Will do.  Thank you.

**Comment by nathanmarz** [comment]
Any update on this? Would really like to get this merged in.

**Comment by d2r** [comment]
I'll take a look again tonight.

**Comment by d2r** [comment]
- Removed debug code
- Restructured validation function as per @nathanmarz suggestions
- Moved validators into a new file so that I could write tests.  Calling getFields on Config was collecting the validators along with the actual config fields & schemas.
- Tests do not cover every config & schema.  I wrote one test to check the validation of TOPOLOGY_WORKERS, since that is what caused #581.
- I did not change when configs were validated, because I did not see a good way to avoid validating the defaults.  One effect is that validation code will run at compile time.
- Regarding Number/Integer, it seems that Number is being coerced to int when we call from clj->java method.
- Fixed a new default that was not added to Config.java
- Unit tests passed.

**Comment by nathanmarz** [comment]
Awesome work. Merged.

File: storm-core/src/jvm/backtype/storm/ConfigValidation.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+            if (o == null) {
+                // A null value is acceptable.
+                return;
+            }
+            final long i;
+            if (o instanceof Number &&
+                    (i = ((Number)o).longValue()) == ((Number)o).doubleValue())
+            {
+                // Test whether the integer is a power of 2.
+                if (i > 0 && (i & (i-1)) == 0) {
```
Comment (xumingming) [inline storm-core/src/jvm/backtype/storm/ConfigValidation.java:75]: very interesting(good) method of validating whether a number is a power of 2!

**Pull request merged**